src/sql/src/catalog.rs

```diff
@@ -1261,24 +1261,26 @@ impl CatalogError {
 
 impl Error for CatalogError {}
 
+// Implements a trait with a generic parameter T and a lifetime 'a for that parameter
+pub(crate) trait SubsourceCatalog<'a, T> {
```

Comment:
You don't need the lifetime here, when something is generic over `T` it means that `T` can become *any* type, including references, which are just another type! This is a common misconception and this is an enlightening read on the topic https://github.com/pretzelhammer/rust-blog/blob/master/posts/common-rust-lifetime-misconceptions.md

I attach the diff that fixes the lifetimes since I tried it out before commenting anyway:

```diff
diff --git a/src/sql/src/catalog.rs b/src/sql/src/catalog.rs
index d099f17e33..3741bad2a1 100644
--- a/src/sql/src/catalog.rs
+++ b/src/sql/src/catalog.rs
@@ -1262,17 +1262,17 @@ impl CatalogError {
 impl Error for CatalogError {}
 
 // Implements a trait with a generic parameter T and a lifetime 'a for that parameter
-pub(crate) trait SubsourceCatalog<'a, T> {
-    fn resolve(&self, item: UnresolvedItemName) -> Result<(UnresolvedItemName, &'a T), PlanError>;
+pub(crate) trait SubsourceCatalog<T> {
+    fn resolve(&self, item: UnresolvedItemName) -> Result<(UnresolvedItemName, &T), PlanError>;
 }
 
 /// Provides a method of generating a 3-layer catalog on the fly, and then
 /// resolving objects within it.
-pub(crate) struct SubsourceCatalogTriple<'a, T>(
-    pub BTreeMap<String, BTreeMap<String, BTreeMap<String, &'a T>>>,
+pub(crate) struct SubsourceCatalogTriple<T>(
+    pub BTreeMap<String, BTreeMap<String, BTreeMap<String, T>>>,
 );
 
-impl<'a, T> SubsourceCatalog<'a, T> for SubsourceCatalogTriple<'a, T> {
+impl<T> SubsourceCatalog<T> for SubsourceCatalogTriple<T> {
     /// Returns the fully qualified name for `item`, as well as the `T` that it
     /// describes.
     ///
@@ -1280,7 +1280,7 @@ impl<'a, T> SubsourceCatalog<'a, T> for SubsourceCatalogTriple<'a, T> {
     /// - If `item` cannot be normalized to a [`PartialItemName`]
     /// - If the normalized `PartialItemName` does not resolve to an item in
     ///   `self.0`.
-    fn resolve(&self, item: UnresolvedItemName) -> Result<(UnresolvedItemName, &'a T), PlanError> {
+    fn resolve(&self, item: UnresolvedItemName) -> Result<(UnresolvedItemName, &T), PlanError> {
         let name = normalize::unresolved_item_name(item)?;
 
         let schemas = match self.0.get(&name.item) {
@@ -1314,7 +1314,7 @@ impl<'a, T> SubsourceCatalog<'a, T> for SubsourceCatalogTriple<'a, T> {
         };
 
         let desc = match databases.get(database) {
-            Some(desc) => *desc,
+            Some(desc) => desc,
             None => sql_bail!("database {database} not found source"),
         };
 
@@ -1332,9 +1332,9 @@ impl<'a, T> SubsourceCatalog<'a, T> for SubsourceCatalogTriple<'a, T> {
 
 /// Provides a method of generating a 2-layer catalog on the fly, and then
 /// resolving objects within it.
-pub(crate) struct SubsourceCatalogDouble<'a, T>(pub BTreeMap<String, BTreeMap<String, &'a T>>);
+pub(crate) struct SubsourceCatalogDouble<T>(pub BTreeMap<String, BTreeMap<String, T>>);
 
-impl<'a, T> SubsourceCatalog<'a, T> for SubsourceCatalogDouble<'a, T> {
+impl<T> SubsourceCatalog<T> for SubsourceCatalogDouble<T> {
     /// Returns the fully qualified name for `item`, as well as the `T` that it
     /// describes.
     ///
@@ -1342,7 +1342,7 @@ impl<'a, T> SubsourceCatalog<'a, T> for SubsourceCatalogDouble<'a, T> {
     /// - If `item` cannot be normalized to a [`PartialItemName`]
     /// - If the normalized `PartialItemName` does not resolve to an item in
     ///   `self.0`.
-    fn resolve(&self, item: UnresolvedItemName) -> Result<(UnresolvedItemName, &'a T), PlanError> {
+    fn resolve(&self, item: UnresolvedItemName) -> Result<(UnresolvedItemName, &T), PlanError> {
         let name = normalize::unresolved_item_name(item)?;
 
         let schemas = match self.0.get(&name.item) {
@@ -1361,7 +1361,7 @@ impl<'a, T> SubsourceCatalog<'a, T> for SubsourceCatalogDouble<'a, T> {
         };
 
         let desc = match schemas.get(schema) {
-            Some(desc) => *desc,
+            Some(desc) => desc,
             None => sql_bail!("schema {schema} not found in source"),
         };
 
diff --git a/src/sql/src/pure.rs b/src/sql/src/pure.rs
index 61a5d33f71..72e38bc506 100644
--- a/src/sql/src/pure.rs
+++ b/src/sql/src/pure.rs
@@ -80,7 +80,7 @@ pub(crate) mod error;
 mod mysql;
 mod postgres;
 
-fn subsource_gen<'a, T, F: SubsourceCatalog<'a, T>>(
+fn subsource_gen<'a, T, F: SubsourceCatalog<&'a T>>(
     selected_subsources: &mut Vec<CreateSourceSubsource<Aug>>,
     catalog: &F,
     source_name: &UnresolvedItemName,
@@ -117,7 +117,7 @@ fn subsource_gen<'a, T, F: SubsourceCatalog<'a, T>>(
 
         let (qualified_upstream_name, desc) = catalog.resolve(subsource.reference.clone())?;
 
-        validated_requested_subsources.push((qualified_upstream_name, subsource_name, desc));
+        validated_requested_subsources.push((qualified_upstream_name, subsource_name, *desc));
     }
 
     Ok(validated_requested_subsources)
diff --git a/src/sql/src/pure/mysql.rs b/src/sql/src/pure/mysql.rs
index a7382c43d2..8bd72b3d34 100644
--- a/src/sql/src/pure/mysql.rs
+++ b/src/sql/src/pure/mysql.rs
@@ -26,7 +26,7 @@ use crate::plan::{PlanError, StatementContext};
 
 pub(super) fn derive_catalog_from_tables<'a>(
     tables: &'a [MySqlTableDesc],
-) -> Result<SubsourceCatalogDouble<'a, MySqlTableDesc>, PlanError> {
+) -> Result<SubsourceCatalogDouble<&'a MySqlTableDesc>, PlanError> {
     // An index from table name -> schema name -> MySqlTableDesc
     let mut tables_by_name = BTreeMap::new();
     for table in tables.iter() {
diff --git a/src/sql/src/pure/postgres.rs b/src/sql/src/pure/postgres.rs
index 366e3cdadc..fab2b37b35 100644
--- a/src/sql/src/pure/postgres.rs
+++ b/src/sql/src/pure/postgres.rs
@@ -33,7 +33,7 @@ use super::error::PgSourcePurificationError;
 pub(super) fn derive_catalog_from_publication_tables<'a>(
     database: &'a str,
     publication_tables: &'a [PostgresTableDesc],
-) -> Result<SubsourceCatalogTriple<'a, PostgresTableDesc>, PlanError> {
+) -> Result<SubsourceCatalogTriple<&'a PostgresTableDesc>, PlanError> {
     // An index from table name -> schema name -> database name -> PostgresTableDesc
     let mut tables_by_name = BTreeMap::new();
     for table in publication_tables.iter() {
@@ -76,7 +76,7 @@ pub(super) async fn validate_requested_subsources_privileges(
 }
 
 pub(super) fn generate_text_columns(
-    publication_catalog: &SubsourceCatalogTriple<'_, PostgresTableDesc>,
+    publication_catalog: &SubsourceCatalogTriple<&PostgresTableDesc>,
     text_columns: &mut [UnresolvedItemName],
     option_name: &str,
 ) -> Result<BTreeMap<u32, BTreeSet<String>>, PlanError> {

```

Reply:
That link is very enlightening! Will refactor with your changes to avoid the unnecessary lifetimes